bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/VerisureSession.java

```diff
@@ -0,0 +1,901 @@
+            "https://m-api02.verisure.com");
+    private int apiServerInUseIndex = 0;
+    private String apiServerInUse = APISERVERLIST.get(apiServerInUseIndex);
+    private boolean areWeLoggedOut = true;
+    private String authstring = "";
+    private @Nullable String csrf;
+    private @Nullable String pinCode;
+    private HttpClient httpClient;
+    private @Nullable String userName = "";
+    private String passwordName = "vid";
+    private @Nullable String password = "";
+
+    public VerisureSession(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public boolean initialize(@Nullable String authstring, @Nullable String pinCode, @Nullable String userName) {
+        logger.debug("VerisureSession:initialize");
+        if (authstring != null) {
+            this.authstring = authstring.substring(0);
+            this.pinCode = pinCode;
+            this.userName = userName;
+            // Try to login to Verisure
+            if (logIn()) {
+                return getInstallations();
+            } else {
+                logger.warn("Failed to login to Verisure!");
+                return false;
+            }
+        }
+        return false;
+    }
+
+    public boolean refresh() {
+        logger.debug("VerisureSession:refresh");
+        int statusCode = areWeLoggedIn();
+        if (statusCode == 503) {
+            return false;
+        } else if (!areWeLoggedOut && (statusCode == 1)) {
+            updateStatus();
+            return true;
+        } else {
+            if (logIn()) {
+                updateStatus();
+                areWeLoggedOut = false;
+                return true;
+            } else {
+                areWeLoggedOut = true;
+                return false;
+            }
+        }
+    }
+
+    public int sendCommand(String url, String data, BigDecimal installationId) {
+        logger.debug("Sending command with URL {} and data {}", url, data);
+        configureInstallationInstance(installationId);
+        int httpResultCode = setSessionCookieAuthLogin();
+        if (httpResultCode == HttpStatus.OK_200) {
+            return postVerisureAPI(url, data);
+        } else {
+            return httpResultCode;
+        }
+    }
+
+    public boolean unregisterDeviceStatusListener(DeviceStatusListener deviceStatusListener) {
+        logger.debug("unregisterDeviceStatusListener for listener {}", deviceStatusListener);
+        return deviceStatusListeners.remove(deviceStatusListener);
+    }
+
+    public boolean registerDeviceStatusListener(DeviceStatusListener deviceStatusListener) {
+        logger.debug("registerDeviceStatusListener for listener {}", deviceStatusListener);
+        return deviceStatusListeners.add(deviceStatusListener);
+    }
+
+    public void dispose() {
+    }
+
+    public @Nullable VerisureThing getVerisureThing(String deviceId) {
+        return verisureThings.get(deviceId);
+    }
```

Comment:
Ugh, you brought up a good point. This thing cache really throws a wrench into safely using generics. The only proper way you could handle it would be to change this method slightly.
```suggestion
    public <T extneds VerisureThing> @Nullable T getVerisureThing(String deviceId, Class<T> thingType) {
        VerisureThing thing = verisureThings.get(deviceId);
        if(thing != null && thingType.isInstance(thing)){
            return (T) thing;
        }
        return null;
    }
```
Although technically you could do generic casting without passing in a Class parameter, it is incredibly unsafe unless you are REALLY sure that it can safely be cast.
I also don't know for sure if `@Nullable` was used correctly in my code above, so if it doesn't work let me know.

Reply:
Could you instead solve it in VTH by doing:

```
T thing = (T) session.getVerisureThing(deviceId);
 if (thing != null) {
         update(thing);
```
In both solutions I a get a compiler warning stating:

```
Type safety: Unchecked cast from VerisureThing to 
 T
```

where the actual cast is done:

```
return (T) thing;
```
 and:

```
T thing = (T) session.getVerisureThing(deviceId);
```

Reply:
The warnings would have to be suppressed in this case since you can't really get around them given that you are storing a bunch of different subclasses in a single map. The compiler doesn't know any better. 

With your solution it would also work, but it wouldn't be the safest approach, and I wouldn't suggest it unless you are really sure that the casting would always work. The approach I gave would make it impossible to get a ClassCastException.

You would need to suppress the warnings either way though. 😕

Reply:
Since VTH calls the getVerisureThing using the key deviceId without knowing the exact thingtype that has that key, how should the call from VTH look like to match the new Class<T> parameter?
`VerisureThing thing = session.getVerisureThing(deviceId);`

```
@Override
    public void handleCommand(ChannelUID channelUID, Command command) {
        logger.debug("VerisureThingHandler handleCommand, channel: {}, command: {}", channelUID, command);
        if (command instanceof RefreshType) {
            Bridge bridge = getBridge();
            if (bridge != null) {
                BridgeHandler bridgeHandler = bridge.getHandler();
                if (bridgeHandler != null) {
                    bridgeHandler.handleCommand(channelUID, command);
                    String deviceId = config.getDeviceId();
                    VerisureSession session = getSession();
                    if (session != null) {
                        VerisureThing thing = session.getVerisureThing(deviceId);
                        if (thing != null) {
                            update(thing);
                        } else {
                            logger.debug("Thing is null!");
                        }
....

```

Reply:
All VTH implement the `getVerisureThingClass()` now, so you can just use it when calling `getVerisureThing`

Reply:
`VerisureThing thing = session.getVerisureThing(deviceId, getVerisureThingClass());`

Reply:
Code looks much cleaner now, really cool to learn from a pro about Generics. 👍 

I took my Java certfication 1999 and haven't actively coded Java for many years so this was really cool to learn :-)

Binding is currently running in my production system and my basic tests have not found any problem, but I'll keep monitoring it.

Reply:
Well I'm happy to hear that, Java has changed a lot since 1999. Although generics become very not fun once wildcards get involved. 🙁